### DISCUSSION TITLE
How to subscribe to current/oldest pointers changes

### DISCUSSION THREAD

**Discussion by ambroseus** [mighdoll/undo-peasy#3] [Q&A]
How to subscribe to  current/oldest pointers changes to check when  undo/redo is allowed?

**Comment by ambroseus** [comment]
working solution with own `undoRedo` middleware 
```js
export function undoRedo(): Middleware {
  const result = (api: MiddlewareAPI) => (next: Dispatch<AnyAction>) => (
    action: AnyAction
  ) => {
    if (alwaysSkipAction(action.type) || undoAction(action.type)) {
      const result = next(action)
      if (undoAction(action.type)) {
        const currentPointer = Number(
          window.localStorage.getItem('undo-redo-state-current')
        )
        const oldestPointer = Number(
          window.localStorage.getItem('undo-redo-state-oldest')
        )

        api.dispatch({
          type: '@action.setUndoState',
          payload: { currentPointer, oldestPointer },
        })
      }
      return result
    } else {
      const result = next(action)
      api.dispatch({ type: '@action.undoSave', payload: action })
      return result
    }
  }

  return result
}

function alwaysSkipAction(actionType: string): boolean {
  return (
    actionType === '@action.ePRS' ||
    actionType === '@@INIT' ||
    actionType === '@action.setUndoState'
  )
}
```

**Reply by mighdoll** [reply]
Nice idea! 

I'd be happy to merge a patch for this feature if you'd like to integrate it.

A few thoughts:
* perhaps extend ```WithUndo``` and ```undoableModelAndHistory``` to include the state about the size of the undo redo stacks. That way users of this feature don't have to do anything.
* I imagine that you want the size of the stacks, rather than the pointers. Is that right? e.g. if the oldest undo state is undo-redo-7 and the current is undo-redo-8, you'd want to report that there is 1 undo available, not 8.  See e.g. the unit test "maxHistory works with redo too".
* Consider capturing the stack sizes in ```HistoryStore``` rather than in the middleware. The HistoryStore encapsulates the knowledge about queue sizes, and so is probably better positioned to know when to adjust the queue sizes.

**Reply by ambroseus** [reply]
@mighdoll  yep)) I'll prepare PR soon..

btw, I also use helper to cleanup localStorage (currently bruteforced), guess it would be helpful to include it in package (smth. like `HistoryStore.cleanup()`)
```js
export function cleanupUndoStorage() {
  let count = 0

  const current = Number(
    window.localStorage.getItem(`undo-redo-state-current`) || 0
  )

  window.localStorage.removeItem(`undo-redo-state-current`)
  window.localStorage.removeItem(`undo-redo-state-oldest`)

  while (count < current + 1) {
    window.localStorage.removeItem(`undo-redo-${count}`)
    count++
  }
}
```

**Reply by ambroseus** [reply]
I also use an additional pointer `newest`

- **undo** is allowed when `current` != `oldest`
- **redo** is allowed when `current` != `newest`
- **on save:**  `newest`++, `current` := `newest`
- **on undo:** `current`-- if `current` > `oldest`
- **on redo:** current++ if `current` < `newest`
